src/v/cloud_storage/remote_segment.cc

```diff
@@ -881,17 +910,91 @@ ss::future<> remote_segment::run_hydrate_bg() {
     _hydration_loop_running = false;
 }
 
-namespace {
+ss::future<> remote_segment::service_chunk_requests() {
+    fragmented_vector<ss::future<ss::file>> chunk_op_results;
+    chunk_op_results.reserve(_chunk_waiters.size());
+
+    fragmented_vector<chunk_request> requests;
+    requests.swap(_chunk_waiters);
+
+    std::ranges::transform(
+      requests,
+      std::back_inserter(chunk_op_results),
+      [this](const auto& request) {
+          vlog(
+            _ctxlog.debug,
+            "Downloading chunk {} with prefetch {}",
+            request.start,
+            request.prefetch);
+          return hydrate_and_materialize_chunk(
+            {request.start, request.prefetch});
+      });
+
+    auto results = co_await ss::when_all(
+      chunk_op_results.begin(), chunk_op_results.end());
+
+    for (size_t i = 0; i < results.size(); ++i) {
+        auto request = std::move(requests[i]);
+        auto current_result = std::move(results[i]);
+
+        ss::file materialized_handle;
+        std::exception_ptr err;
 
+        if (current_result.failed()) {
+            err = current_result.get_exception();
+        } else {
+            materialized_handle = current_result.get();
+        }
+
+        // Materialization may fail due to cache eviction. In this case the
+        // materialized handle is default initialized. Re-queue waiter and try
+        // again later. Continue with other requests in queue.
```

Comment:
> try again later

What does later mean here? Next time a chunk is downloaded? Should we be hitting the condition variable again if we requeue?

Reply:
Fixed the comment, later here means on the next loop iteration in `run_hydrate_bg`. The condition variable has a predicate which uses `do_until`, so it should avoid the call to wait since the predicate is true on the next loop iteration.